File: backend/app/routers/sets.py
"""Flashcard Set API endpoints for creation, listing, searching, moving, and updating."""

from typing import Annotated

from fastapi import APIRouter, Depends, Query, Response, status

from app.dependencies import get_current_user_id, get_set_service
from app.models.set import SetCreate, SetResponse, SetUpdate
from app.services.set_service import SetService

router = APIRouter(prefix="/api/v1/sets", tags=["Sets"])


@router.post(
    "",
    response_model=SetResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a flashcard set",
)
async def create_set(
    data: SetCreate,
    user_id: Annotated[str, Depends(get_current_user_id)],
    set_service: Annotated[SetService, Depends(get_set_service)],
) -> SetResponse:
    """Create a new flashcard set within an optional folder."""
    return await set_service.create_set(data, user_id)


@router.get(
    "",
    response_model=list[SetResponse],
    summary="List flashcard sets",
)
async def list_sets(
    user_id: Annotated[str, Depends(get_current_user_id)],
    set_service: Annotated[SetService, Depends(get_set_service)],
    folder_id: str | None = Query(default=None, description="Filter sets by containing folder ID."),
    root_only: bool = Query(default=False, description="Filter sets located at the root level."),
    tag: str | None = Query(default=None, description="Filter sets matching a specific tag."),
    search: str | None = Query(default=None, description="Search sets by name or description."),
) -> list[SetResponse]:
    """List and filter flashcard sets for the current user."""
    return await set_service.list_sets(
        user_id=user_id,
        folder_id=folder_id,
        root_only=root_only,
        tag=tag,
        search=search,
    )


@router.get(
    "/{set_id}",
    response_model=SetResponse,
    summary="Get flashcard set by ID",
)
async def get_set(
    set_id: str,
    user_id: Annotated[str, Depends(get_current_user_id)],
    set_service: Annotated[SetService, Depends(get_set_service)],
) -> SetResponse:
    """Retrieve full details of a specific flashcard set."""
    return await set_service.get_set(set_id, user_id)


@router.patch(
    "/{set_id}",
    response_model=SetResponse,
    summary="Update flashcard set",
)
async def update_set(
    set_id: str,
    data: SetUpdate,
    user_id: Annotated[str, Depends(get_current_user_id)],
    set_service: Annotated[SetService, Depends(get_set_service)],
) -> SetResponse:
    """Update set title, description, tags, or move to another folder."""
    return await set_service.update_set(set_id, data, user_id)


@router.delete(
    "/{set_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete flashcard set",
)
async def delete_set(
    set_id: str,
    user_id: Annotated[str, Depends(get_current_user_id)],
    set_service: Annotated[SetService, Depends(get_set_service)],
) -> Response:
    """Delete a flashcard set and all cards contained within it."""
    await set_service.delete_set(set_id, user_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: mcp/src/tools/sets.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { EasyFlashcardApiClient } from '../client/apiClient.js';

export function registerSetTools(server: McpServer, client: EasyFlashcardApiClient): void {
  // 1. List Sets
  server.tool(
    'list_sets',
    'List flashcard sets in the user workspace with summary information (card count, tags, description).',
    {
      folder_id: z
        .string()
        .optional()
        .describe('Filter sets contained in a specific folder ID.'),
      root_only: z
        .boolean()
        .optional()
        .describe('If true, returns only sets at the root workspace level.'),
      tag: z
        .string()
        .optional()
        .describe('Filter sets matching a specific tag.'),
      search: z
        .string()
        .optional()
        .describe('Search sets by name or description.'),
    },
    async (params) => {
      const sets = await client.listSets({
        folderId: params.folder_id,
        rootOnly: params.root_only,
        tag: params.tag,
        search: params.search,
      });

      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(sets, null, 2),
          },
        ],
      };
    }
  );

  // 2. Get Set (with optional cards)
  server.tool(
    'get_set',
    'Get full details of a specific flashcard set, including all cards contained within it for easy studying or quiz generation.',
    {
      set_id: z.string().describe('Unique ID of the flashcard set.'),
      include_cards: z
        .boolean()
        .optional()
        .default(true)
        .describe('If true (default), returns all flashcards in the set alongside the set metadata.'),
    },
    async ({ set_id, include_cards }) => {
      if (include_cards !== false) {
        const setWithCards = await client.getSetWithCards(set_id);
        return {
          content: [
            {
              type: 'text',
              text: JSON.stringify(setWithCards, null, 2),
            },
          ],
        };
      }

      const setInfo = await client.getSet(set_id);
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(setInfo, null, 2),
          },
        ],
      };
    }
  );

  // 3. Create Set
  server.tool(
    'create_set',
    'Create a new flashcard set within an optional folder.',
    {
      name: z.string().min(1).describe('Title of the flashcard set (e.g. "Photosynthesis & Cellular Respiration").'),
      folder_id: z
        .string()
        .optional()
        .describe('Optional ID of the folder to place this set inside. Leave empty for root workspace.'),
      description: z
        .string()
        .optional()
        .describe('Optional description or overview of the topics covered in this set.'),
      tags: z
        .array(z.string())
        .optional()
        .describe('Optional categorization tags (e.g. ["Biology", "ExamPrep"]).'),
    },
    async (params) => {
      const newSet = await client.createSet({
        name: params.name,
        folderId: params.folder_id ?? null,
        description: params.description ?? null,
        tags: params.tags ?? [],
      });

      return {
        content: [
          {
            type: 'text',
            text: `Successfully created flashcard set "${newSet.name}" (ID: ${newSet.id}).\n\n${JSON.stringify(
              newSet,
              null,
              2
            )}`,
          },
        ],
      };
    }
  );

  // 4. Update Set
  server.tool(
    'update_set',
    'Update a flashcard set title, description, tags, or move it to a different folder.',
    {
      set_id: z.string().describe('ID of the flashcard set to update.'),
      name: z.string().optional().describe('New title for the set.'),
      folder_id: z
        .string()
        .optional()
        .describe('New folder ID to move this set into (or null/empty for root).'),
      description: z.string().optional().describe('New description for the set.'),
      tags: z.array(z.string()).optional().describe('New list of tags for the set.'),
    },
    async ({ set_id, name, folder_id, description, tags }) => {
      const updatedSet = await client.updateSet(set_id, {
        name,
        folderId: folder_id,
        description,
        tags,
      });

      return {
        content: [
          {
            type: 'text',
            text: `Successfully updated flashcard set "${updatedSet.name}" (ID: ${updatedSet.id}).\n\n${JSON.stringify(
              updatedSet,
              null,
              2
            )}`,
          },
        ],
      };
    }
  );

  // 5. Delete Set
  server.tool(
    'delete_set',
    'Delete a flashcard set and all cards contained within it.',
    {
      set_id: z.string().describe('ID of the flashcard set to delete.'),
    },
    async ({ set_id }) => {
      await client.deleteSet(set_id);
      return {
        content: [
          {
            type: 'text',
            text: `Successfully deleted flashcard set ${set_id} and all its cards.`,
          },
        ],
      };
    }
  );
}
